### PR TITLE
Add examples for basic profile.

### PR DESCRIPTION
**Pull request by SunFulong**
I'm going to add examples for each sections in basic profile, here is a preview for session section.

Should I continue this job?

PS: Removed some bad white spaces.

### CODE REVIEW

**Comment by ecorm** [comment]
If it is decided that such lengthy examples are to be included, I suggest that they be added in a appendix to reduce clutter in the "reference" section of the standard.

**Comment by SunFulong** [comment]
> I suggest that they be added in an appendix to reduce clutter in the "normative" section of the standard.

OK, I'll move to a separate chapter at the end of basic profile.

Let's discuss should I do it and how to do it?

**Comment by oberstet** [comment]
> I'm going to add examples for each sections in basic profile

What issue is fixed with this PR?

**Comment by SunFulong** [comment]
> What issue is fixed with this PR?

No fixes in this PR, just add examples to help developers understand each progress.

**Comment by SunFulong** [comment]
> You probably would not want to read lengthy message conversations plain text.

Around me, there are many developers who have difficulty reading and understanding documents (especially documents in non-native languages), no matter how long or short they are.

But, a detailed example can help to understand.

**Comment by oberstet** [comment]
> Around me, there are many developers who have difficulty reading and understanding documents (especially documents in non-native languages), no matter how long or short they are.

ok, I see. interesting. for me, it's like this: the longer a document, the more unlikely I am going to invest any time in it. because I assume that either the authors didn't take their time to condense to the essence or it's bullshit anyways. reading such material is usually a waste of time for me. however, I can see this is only one perspective. as said, I'd be +1 on adding machine readable full message conversations as extra files and list/shortly describe those in an appendix. I am -1 on adding message flows to the spec text in-line, and also -1 on adding extended message flows as non-machine readable "human language text".

**Comment by SunFulong** [comment]
You mentioned "machine readable message" many times, how should I write it? Could you show me?

**Comment by oberstet** [comment]
> You mentioned "machine readable message" many times, how should I write it? Could you show me?

Ok, I will write an example .. a proposal that we can then discuss for a common format we could use. Will take me an hour or so, I'll post here

**Comment by oberstet** [comment]
grafik

**Comment by SunFulong** [comment]
I'm even more confused, and now I don't know where to start!😭😭

**Comment by oberstet** [comment]
@SunFulong no worries! let me give you another example, here is a complete trace of all sent and received messages on 2 WAMP sessions doing basic PubSub:

https://gist.github.com/oberstet/31c7e4f187678fed103e5ba79d4b4266

**Comment by oberstet** [comment]
for example, the very first message is HELLO sent by session 1:

```
[
  "TX",
  "realm1",
  null,
  [
    1,
    "realm1",
    {
      "roles": {
        "caller": {
          "features": {
            "caller_identification": true,
            "call_canceling": true,
            "progressive_call_results": true
          }
        },
        "callee": {
          "features": {
            "caller_identification": true,
            "pattern_based_registration": true,
            "shared_registration": true,
            "progressive_call_results": true,
            "registration_revocation": true
          }
        },
        "publisher": {
          "features": {
            "publisher_identification": true,
            "subscriber_blackwhite_listing": true,
            "publisher_exclusion": true
          }
        },
        "subscriber": {
          "features": {
            "publisher_identification": true,
            "pattern_based_subscription": true,
            "subscription_revocation": true
          }
        }
      }
    }
  ]
]
```

That is, the logging is as an array of:

* TX or RX
* realm
* session ID
* array with WAMP message

**Comment by SunFulong** [comment]
Ya, I see.

So, I need write each story in a dedicated file, then put the files in a new folder?

How to organize the file structure?

**Comment by oberstet** [comment]
> So, I need write each story in a dedicated file, then put the files in a new folder?

yeah, I like that: one "story" == one multi-session message conversation == one (JSON) file

and yes, a new folder like `wamp-proto/rfc/traces` with files like `wamp-proto/rfc/traces/pubsub_basis.json` ..

**Comment by oberstet** [comment]
> So, I need write each story in a dedicated file, then put the files in a new folder?

yes, I like that: one "story" == one message conversation log (with messages from one or more sessions), and in a new folder like `wamp-proto/rfc/traces/pubsub_basic1.json` ...

**Comment by oberstet** [comment]
ok, you can find trace files here:

https://github.com/crossbario/autobahn-js/blob/master/packages/autobahn/test/test_pubsub_basic.trace

> those files have each message separated by a newline (with nothing else), so it's not json in itself .. but can be parsed streaming/easy

there are files for every test in autobahn-js:

```
(cpy311_1) (base) oberstet@intel-nuci7:~/scm/crossbario/autobahn-js$ find packages/autobahn/test/*.trace
packages/autobahn/test/test_connect.trace
packages/autobahn/test/test_error_handling.trace
packages/autobahn/test/test_pubsub_basic.trace
packages/autobahn/test/test_pubsub_complex.trace
packages/autobahn/test/test_pubsub_eligible.trace
packages/autobahn/test/test_pubsub_excludeme.trace
packages/autobahn/test/test_pubsub_exclude.trace
packages/autobahn/test/test_pubsub_multiple_matching_subs.trace
packages/autobahn/test/test_pubsub_options.trace
packages/autobahn/test/test_pubsub_prefix_sub.trace
packages/autobahn/test/test_pubsub_wildcard_sub.trace
packages/autobahn/test/test_rawsocket_transport.trace
packages/autobahn/test/test_rpc_arguments.trace
packages/autobahn/test/test_rpc_complex.trace
packages/autobahn/test/test_rpc_error.trace
packages/autobahn/test/test_rpc_options.trace
packages/autobahn/test/test_rpc_progress.trace
packages/autobahn/test/test_rpc_request_id_sequence.trace
packages/autobahn/test/test_rpc_routing.trace
packages/autobahn/test/test_rpc_slowsquare.trace
packages/autobahn/test/test_serialization_cbor.trace
packages/autobahn/test/test_serialization_json.trace
```

**Comment by SunFulong** [comment]
> yeah, I like that: one "story" == one multi-session message conversation == one (JSON) file
>
> and yes, a new folder like `wamp-proto/rfc/traces` with files like `wamp-proto/rfc/traces/pubsub_basis.json` ..

Two questions:

1. A story may have 4 roles: SubscriberA, SubscriberB, Publisher, Broker, should they be placed in 4 files?
2. How can I explain the story details to human readers?

**Comment by oberstet** [comment]
> 1. A story may have 4 roles: SubscriberA, SubscriberB, Publisher, Broker, should they be placed in 4 files?

With the approach I examplified, there are only *Client* roles as the message traces are from the client's perspective.

There can be 1-many client session each with one or more of the 4 client roles.

Regardless of the number of client sessions, all messages of all (client) sessions (sent or received) are logged in one file.

The above tests run all client sessions from one client program with multiple client-router connections (one per session). The order of messages in the combined message log is "exactly as happened" for that 1 client program, and written as such.

Currently, the traces do _not_ contain:

* router URL (as all client session connect to the same one router node)
* client side timestamps or cross-session client-program-side sequence IDs

*We might want to add those in the future.*

> How can I explain the story details to human readers?

Consider this together with the respective trace log

https://github.com/crossbario/autobahn-js/blob/master/packages/autobahn/test/test_pubsub_multiple_matching_subs.js

This opens 5 sessions and does:

* Client A: Publish messages to `com.myapp.topic1` (with publish acknowledge, but without self-subscribing)
* Client B: Subscribe to `com..topic1` with wildcard matching
* Client C: Subscribe to `com.myapp` with prefix matching
* Client D: Subscribe to `com` with prefix matching
* Client E: Subscribe to `com.myapp.topic1` with exact matching

The example demonstrates PubSub with events matching multiple patterns (subscriptions).

**Comment by ecorm** [comment]
In CppWAMP, to make my traces more readable, I combine a string with the message type number like this:

```
[
    {"PUBLISH": 16},
    3,
    {},
    "com.myapp.topic1",
    [
      2
    ]
]
```

A tuple could also work:
```
[
    [16, "PUBLISH"],
    3,
    {},
    "com.myapp.topic1",
    [
      2
    ]
]
```

It does not match exactly what's sent over the wire, but it avoids having to memorize/lookup a bunch of message type numbers.

**Comment by SunFulong** [comment]
I can understand the example you are showing.

I mean: I can put .trace files in that folder, but our spec readers (developers), how can they understand the trace files? I put a text or markdow file in the same folder?

**Comment by oberstet** [comment]
> I mean: I can put .trace files in that folder, but our spec readers (developers), how can they understand the trace files? I put a text or markdow file in the same folder?

aah, ok, sorry, didn't get the question first.

my suggestion would be:

1. add plain/raw files to a new folder as indicated above
2. add a new appendix section for the spec, eg a file `wamp-proto/rfc/text/traces.md` and add that "somehow" as appendix (in backmatter.md? not sure)

in that new text section, have 1 subsection per story / trace file, explaining and cross-referencing. PLUS: _reference_ to the actual trace file ... not sure how this is done "best"

also the "somehow": needs figuring out how to best do so that the [built](https://github.com/wamp-proto/wamp-proto/blob/17f0fc90aa37cbddae7b97c17cff1444a63c738e/Makefile#L85) specs look like expected.

@ecorm any opinions? I mean in general rgd above proposal, and specifically rgd text integration?

**Comment by oberstet** [comment]
> I combine a string with the message type number like this:

ah yes, we could use WAMP message names rather than message type codes. mapping is obviously [easy](https://github.com/crossbario/autobahn-js/blob/master/packages/autobahn/lib/session.js#L190) and makes the resulting log more human readable.

however: I can't run / test those traces identically to a raw message trace from a WAMP transport - which is the nice thing with using exactly the WAMP message as it appeared on the wire, just embedded into those trace records.

if we want _both_, then I'd argue: keep the embedded WAMP message "as is" (including message type code), and add the WAMP message name to the outer enclosing record:

```
["RX"/"TX", <realm>, <session_id>, <msg_type_name>, <WAMP message>]
```

**Comment by ecorm** [comment]
> @ecorm any opinions? I mean in general rgd above proposal, and specifically rgd text integration?

If you're asking about the need for this, I might not be the best person to ask since I already have a fairly deep understanding of the protocol. I guess the cost vs benefit is the cost to write and maintain the examples versus the help it provides to users and implementers. You could check if other competing protocols offer similar "stories" and examples.

The only thing that concerns me is very long examples cluttering up the normative part of the spec, and I see we're already on the same page on that.

> if we want both, then I'd argue: keep the embedded WAMP message "as is" (including message type code), and add the WAMP message name to the outer enclosing record

I like your idea of enclosing to preserve the original message, and I think I'll change the tracing mechanism in CppWAMP to do just that.

**Comment by oberstet** [comment]
one note rgd the traces from ABJS I posted above: those are all ending with all the session sending a (client initiated) GOODBYE:

```
[
  "TX",
  "realm1",
  110043672210347,
  [
    6,
    {},
    "wamp.close.normal"
  ]
]
```

However, I am missing the GOODBYE sent by a router during a fully orderly closing handshake.

The reasons might be the debug logging in ABJS or it may be crossbar being run with hardstyle session closing, just dropping without bothering to reply with GOODBYE first. I haven't looked into this though, just wanted to note in case @SunFulong is wondering ..

**Comment by KSDaemon** [comment]
I think a sequence diagram would be ideal for flow description. Something like that:
```mermaid
sequenceDiagram
    autonumber
    participant C as Client
    participant R as Router
    Note over C, R: TPC/Socket/Websocket Connection Established
    C->>R: Client Hello Message [HELLO, Realm|uri, Details|dict] <br>[1, "org.example", {"agent": "WAMP/2.0 (Client)", "roles": {"publisher": {}, "subscriber": {}}}]
    R->>C: Router Welcome Message [WELCOME, Session|id, Details|dict]<br>[2, 5606911429535183, {"agent": "WAMP/2.0 (Router)", "roles": {"broker": {}}}]
    C->>R: Client Hello Again Message [HELLO, Realm|uri, Details|dict] (Bad) <br>[1, "org.example", {"agent": "WAMP/2.0 (Client)", "roles": {"publisher": {}, "subscriber": {}}}]
    R->>C: Router Abort Message [ABORT, Details|dict, Reason|uri]<br>[3, {"message": "Received HELLO message after session was established."}, "wamp.error.protocol_violation"]
    Note over C, R: Connection Disconnected
```

But it is hard to export mermaid to ascii :(((
btw, you can +1 here https://github.com/mermaidjs/mermaid-live-editor/issues/36

And writing it by hands - well, not much fun %)

File: rfc/text/basic/bp_04_sessions.md [hunk 1]
```diff
@@ -333,6 +333,145 @@ and the other peer replies
...
+
+(Client to Router) Client Hello
+
+{align="left"}
+    [1, "org.example", {"agent": "WAMP/2.0 (Client)", "roles": {"publisher": {}, "subscriber": {}}}]
+
+(Router to Client) Router Welcome
+
+{align="left"}
+    [2, 5606911429535183, {"agent": "WAMP/2.0 (Router)", "roles": {"broker": {}}}]
```
Comment (KSDaemon) [inline rfc/text/basic/bp_04_sessions.md:389]: Regarding schema I mean something like that:

```suggestion
    // [WELCOME, Session|id, Details|dict]
    [2, 5606911429535183, {"agent": "WAMP/2.0 (Router)", "roles": {"broker": {}}}]
```

File: rfc/text/basic/bp_04_sessions.md [hunk 2]
```diff
@@ -333,6 +333,145 @@ and the other peer replies
...
+The differences between `ABORT` and `GOODBYE` messages is that `ABORT` is never replied to by a Peer, whereas `GOODBYE` must be replied to by the receiving Peer.
 
 > Though `ABORT` and `GOODBYE` are structurally identical, using different message types serves to reduce overloaded meaning of messages and simplify message handling code.
+
+
+## All Together
+
+### Correct Exapmle
+
+Here's a complete example showing how a client establishes and closes a session.
```
[the remark this answers is not in the document]
Comment (SunFulong) [reply rfc/text/basic/bp_04_sessions.md:345]: Yes, in this example, what I want to reflect is the need to set the correct "subprotocol" header.

File: rfc/text/basic/bp_04_sessions.md (hunk 1 above)
Comment (SunFulong) [reply rfc/text/basic/bp_04_sessions.md:389]: Right, I'll do it, but I've no chance to do it.

**Comment by oberstet** [comment]
> I think a sequence diagram would be ideal for flow description. Something like that:
> 
> But it is hard to export mermaid to ascii :((( btw, you can +1 here [mermaidjs/mermaid-live-editor#36](https://github.com/mermaidjs/mermaid-live-editor/issues/36)
> 
> And writing it by hands - well, not much fun %)

ah yeah, I think you mentioned that one, looks quite neat!

creating mermaid sequence charts "manually": well, not really scalable for our spec purposes here IMO

BUT: it seems like mermaid supports markdown: https://github.blog/2022-02-14-include-diagrams-markdown-files-mermaid/#enter-mermaid

can I create a sequence diagram as you posted in mermaid-markdown?

does github support that markdown flavor?

**Comment by oberstet** [comment]
indeed, github supports pre code blocks of type "mermaid":


```mermaid
sequenceDiagram
    participant dotcom
    participant iframe
    participant viewscreen
    dotcom->>iframe: loads html w/ iframe url
    iframe->>viewscreen: request template
    viewscreen->>iframe: html & javascript
    iframe->>dotcom: iframe ready
    dotcom->>iframe: set mermaid data on iframe
    iframe->>iframe: render mermaid
```

-----------

```mermaid
  flowchart LR;
      A[CI MULTI CHAPTCHA]-->B{Select captcha service by developer?};
      classDef green color:#022e1f,fill:#00f500;
      classDef red color:#022e1f,fill:#f11111;
      classDef white color:#022e1f,fill:#fff;
      classDef black color:#fff,fill:#000;
      B--YES-->C[How to use?]:::green;
      
      C-->U[I choose recaptcha.]:::green;
      U--Views-->Q["echo CIMC_JS('recaptcha');\n echo CIMC_HTML(['captcha_name'=>'recaptcha']);"]:::green;
      U--Controller-->W["CIMC_RULE('recaptcha');"]:::green;
      
      C-->I[I choose arcaptcha.]:::white;
      I--Views-->O["echo CIMC_JS('arcaptcha');\n echo CIMC_HTML(['captcha_name'=>'arcaptcha']);"]:::white;
      I--Controller-->P["CIMC_RULE('arcaptcha');"]:::white;
      
      C-->X[I choose bibot.]:::red;
      X--Views-->V["echo CIMC_JS('bibot');\n echo CIMC_HTML(['captcha_name'=>'bibot']);"]:::red;
      X--Controller-->N["CIMC_RULE('bibot');"]:::red;
      
      B--NO-->D[How to use?]:::black;
      D---Views:::black-->F["echo CIMC_JS('randomcaptcha');\n echo CIMC_HTML(['captcha_name'=>'randomcaptcha']);"]:::black; 
      D---Controller:::black-->T["CIMC_RULE('archaptcha,recaptcha,bibot');"]:::black;
```

--------

```mermaid
sequenceDiagram
    participant Alice
    participant John
    links Alice: {"Dashboard": "https://dashboard.contoso.com/alice", "Wiki": "https://wiki.contoso.com/alice"}
    links John: {"Dashboard": "https://dashboard.contoso.com/john", "Wiki": "https://wiki.contoso.com/john"}
    Alice->>John: Hello John, how are you?
    John-->>Alice: Great!
    Alice-)John: See you later!
```

**Comment by oberstet** [comment]
here is an example of how we could use mermaid for wamp messaging

```mermaid
sequenceDiagram
loop RPC "add2"
    Caller->>Router: CALL
    Router->>Callee: INVOCATION
    Callee-->>Router: YIELD
    Router-->>Caller: RESULT
end
```

**Comment by KSDaemon** [comment]
@oberstet you got it! :)

**Comment by oberstet** [comment]
> you got it! :)

the only problem with even above figure, the grouping of the 4 messages for the 1 RPC, isn't available via client-side only tracking.

the mapping from procedure name to invocation based on registration by the callee is only available as part of the routing table used by the router internally

could we define and describe "routing table" as necessarily required by a router, but describe at the spec-level only using abstract/logical descriptions?

**Comment by SunFulong** [comment]
Will we move to mermaid? Can it compile to ASCII?

**Comment by KSDaemon** [comment]
> Will we move to mermaid? Can it compile to ASCII?

Unfortunately, I think `no` :( To use mermaid we need to figure out how to embed it into the resulting IETF Specification (I think for now the only way is to export all mermaid diagrams to images and use them). Because in the end we want to have a standard EITF-compliant document, that once become a real RFC :)

**Comment by oberstet** [comment]
@KSDaemon do you know https://bestchai.bitbucket.io/shiviz/?

**Comment by KSDaemon** [comment]
Checked the https://bestchai.bitbucket.io/shiviz/. Well, not so neat as a mermaid :) 
But again: we need a tool that can generate a graph image from the sources. I didn't find info about that. 
Also it seems that this tool is mostly (but not limited to) for analyzing distributed logs in an automatic fashion.

**Comment by oberstet** [comment]
> the cost to write and maintain the examples versus the help it provides to users and implementers

I agree. if the detailed message example traces are not automatically tested/verified, they will have bugs. and this is worse than no example trace.

ABJS is now additionally writing `*.trace` files from automated CI tests

https://github.com/crossbario/autobahn-js/tree/master/packages/autobahn/test

-----

if we want to specify above trace format in the spec, we could add it in a follow up PR

if so, we might even add some trace files to the repo here ... but then the trace files aren't verified/generated.

**Pull request closed without merging**